## Student Performance Indicator

### Life cycle of Machine learning Project

- Understanding the Problem Statement
- Data Collection
- Data Checks to perform
- Exploratory data analysis
- Data Pre-Processing
- Model Training
- Choose best model

### 1) Problem statement
- This project understands how the student's performance (test scores) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.


### 2) Data Collection
- Dataset Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977
- The data consists of 8 column and 1000 rows.

### 2.1 Import Data and Required Packages
####  Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.

In [19]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline 
import warnings
warnings.filterwarnings('ignore')

#### Import the CSV Data as Pandas DataFrame

In [20]:
df = pd.read_csv('data/StudentsPerformance.csv')

#### Show Top 5 Records

In [21]:
df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75


#### Shape of the dataset

In [22]:
df.shape

(1000, 8)

### 2.2 Dataset information

- gender : sex of students  -> (Male/female)
- race/ethnicity : ethnicity of students -> (Group A, B,C, D,E)
- parental level of education : parents' final education ->(bachelor's degree,some college,master's degree,associate's degree,high school)
- lunch : having lunch before test (standard or free/reduced) 
- test preparation course : complete or not complete before test
- math score
- reading score
- writing score

### 3. Data Checks to perform

- 3.1. Check Missing values
- 3.2. Check Duplicates
- 3.3. Check data type
- 3.4. Check the number of unique values of each column
- 3.5. Check statistics of data set
- 3.6. Check various categories present in the different categorical column
- 3.7. Adding Columns

### 3.1 Check Missing values

In [23]:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

#### There are no missing values in the data set

### 3.2 Check Duplicates

In [24]:
df.duplicated().sum()

np.int64(0)

#### There are no duplicates  values in the data set

### 3.3 Check data types

In [25]:
# Check Null and Dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


### 3.4 Checking the number of unique values of each column

In [26]:
# get the number of unique values in each column
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

### 3.5 Check statistics of data set

In [27]:
df.describe()

,math score,reading score,writing score
count,1000.00000,1000.000000,1000.000000
mean,66.08900,69.169000,68.054000
std,15.16308,14.600192,15.195657
min,0.00000,17.000000,10.000000
25%,57.00000,59.000000,57.750000
50%,66.00000,70.000000,69.000000
75%,77.00000,79.000000,79.000000
max,100.00000,100.000000,100.000000


#### Insight
- From above description of numerical data, all means are very close to each other - between 66 and 69.2;
- All standard deviations are also close - between 14.6 and 15.19;
- While there is a minimum score  0 for math, for writing minimum is much higher = 10 and for reading myet higher = 17

### 3.6 Exploring Data (categories present in the different categorical column)

In [28]:
df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75


In [29]:
class EDAAnalyzer:
    def __init__(self,df:pd.DataFrame):
        self.df = df    # Store the reference to the original DataFrame
        
    def get_categorical_columns(self):
        """Return a list of column names that are categorical."""
        cat_cols = self.df.select_dtypes(include='object').columns.tolist()
        return cat_cols
    
    def get_numerical_columns(self):
        """Return a list of column names that are numerical."""
        num_cols = self.df.select_dtypes(exclude='object').columns.tolist()
        return num_cols
    
    def get_categorical_columns_unique_values(self):
        cat_col_unique_vals = dict()
        cat_cols = self.get_categorical_columns()
        for col in cat_cols:
            unique_vals = self.df[col].unique().tolist()
            cat_col_unique_vals[col] = unique_vals
        return cat_col_unique_vals

    def get_categorical_columns_unique_value_dist(self):
        cat_cols = self.get_categorical_columns()
        cat_summary = {col : list(self.df[col].value_counts().items()) for col in cat_cols}
        return cat_summary
    
    def print_cat_cols_unique_vals(self):
        cat_col_unique_vals = self.get_categorical_columns_unique_values()
        print("Categorical Columns : Unique Values in them")
        for cat in cat_col_unique_vals:
            print("Categories in ",cat," column :   ", cat_col_unique_vals[cat])
 

    def get_num_condition(self,feature:str, condition)->int:
        """
        Count number of rows where a condition on a feature is satisfied.

        Parameters:
        - feature (str): Column name
        - condition (Callable): A function/lambda applied to the column values

        Returns:
        - int: Number of rows satisfying the condition
        """
        if feature not in self.df.columns:
            raise ValueError(f"Column '{feature}' not found in DataFrame")
            
        mask = condition(self.df[feature])
        return mask.sum()

In [30]:
eda_analysis = EDAAnalyzer(df)

In [31]:
# get categorical columns unique values:
eda_analysis.print_cat_cols_unique_vals()

Categorical Columns : Unique Values in them
Categories in  gender  column :    ['female', 'male']
Categories in  race/ethnicity  column :    ['group B', 'group C', 'group A', 'group D', 'group E']
Categories in  parental level of education  column :    ["bachelor's degree", 'some college', "master's degree", "associate's degree", 'high school', 'some high school']
Categories in  lunch  column :    ['standard', 'free/reduced']
Categories in  test preparation course  column :    ['none', 'completed']


In [32]:
# define numerical & categorical columns
categorical_features = eda_analysis.get_categorical_columns()
numerical_features = eda_analysis.get_numerical_columns()

# print the features
print('We have {} numerical features : {}'.format(len(numerical_features), numerical_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

We have 3 numerical features : ['math score', 'reading score', 'writing score']

We have 5 categorical features : ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course']


In [33]:
eda_analysis.get_categorical_columns_unique_value_dist()

{'gender': [('female', 518), ('male', 482)],
 'race/ethnicity': [('group C', 319),
  ('group D', 262),
  ('group B', 190),
  ('group E', 140),
  ('group A', 89)],
 'parental level of education': [('some college', 226),
  ("associate's degree", 222),
  ('high school', 196),
  ('some high school', 179),
  ("bachelor's degree", 118),
  ("master's degree", 59)],
 'lunch': [('standard', 645), ('free/reduced', 355)],
 'test preparation course': [('none', 642), ('completed', 358)]}

### 3.7 Adding columns for "Total Score" and "Average"

In [34]:
df['total score'] = df['math score'] + df['reading score'] + df['writing score']
df['average'] = df['total score']/3
df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score,total score,average
0,female,group B,bachelor's degree,standard,none,72,72,74,218,72.666667
1,female,group C,some college,standard,completed,69,90,88,247,82.333333
2,female,group B,master's degree,standard,none,90,95,93,278,92.666667
3,male,group A,associate's degree,free/reduced,none,47,57,44,148,49.333333
4,male,group C,some college,standard,none,76,78,75,229,76.333333


In [35]:
# Numerical features : test scores : condition - full score(100)
reading_full = eda_analysis.get_num_condition('reading score',lambda x:x==100)
writing_full = eda_analysis.get_num_condition('writing score',lambda x:x==100)
math_full = eda_analysis.get_num_condition('math score',lambda x:x==100)


# print the nums
print(f'Number of students with full marks in Maths: {math_full}')
print(f'Number of students with full marks in Writing: {writing_full}')
print(f'Number of students with full marks in Reading: {reading_full}')

Number of students with full marks in Maths: 7
Number of students with full marks in Writing: 14
Number of students with full marks in Reading: 17


In [36]:
# Numerical features : test scores : condition - (<=20)
reading_less_20 = eda_analysis.get_num_condition('reading score', lambda x:x<=20)
writing_less_20 = eda_analysis.get_num_condition('writing score', lambda x:x<=20)
math_less_20 = eda_analysis.get_num_condition('math score', lambda x:x<=20)

print(f'Number of students with less than 20 marks in Maths: {math_less_20}')
print(f'Number of students with less than 20 marks in Writing: {writing_less_20}')
print(f'Number of students with less than 20 marks in Reading: {reading_less_20}')

Number of students with less than 20 marks in Maths: 4
Number of students with less than 20 marks in Writing: 3
Number of students with less than 20 marks in Reading: 1


#####  Insights
 - From above values we get students have performed the worst in Maths 
 - Best performance is in reading section